### PR TITLE
Addresses discourse username becoming out of sync when Drupal username changes

### PR DESCRIPTION
**Pull request by aaronbauman**
Background on this issue in https://github.com/United-Philanthropy-Forum/km-collaborative/issues/1110

Includes:

* New API endpoint getUsers to fetch a list of users from Discourse
* hook_discourse_sso_parameters_alter implementation to keep Discourse username in sync on the Drupal side when a user logs in via SSO
* A drush script to clean up existing out-of-sync records.

### CODE REVIEW

File: fix-discourse-usernames.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  if (empty($discourse_data)) {
+    break;
+  }
+  foreach ($discourse_data as $user) {
+    $discourse_cleaned[$user->id] = $user;
+  }
+  $page++;
+
+  $query = \Drupal::entityTypeManager()->getStorage('user')->getQuery();
+  $query->condition('discourse_user_field__user_id', array_keys($discourse_cleaned), 'IN');
```
Comment (aaronbauman) [inline fix-discourse-usernames.php:32]: Make sure there aren't any users with duplicate Discourse ids

**Comment by aaronbauman** [comment]
tagging @unclegcb for review

Michigan is running with this patch for the past couple months. Haven't heard any feedback from Will or others yet.

File: fix-discourse-usernames.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        ->notice('Skip empty account :uid', [':uid' => $uid]);
+      continue;
+    }
+    $discourse_id = $account->discourse_user_field->user_id;
+    if (empty($discourse_cleaned[$discourse_id])) {
+      \Drupal::logger('discourse')
+        ->notice('No discourse match found for user :uid username from discourse :username', [
+          ':uid' => $account->id(),
+          ':username' => $account->getAccountName()
+        ]);
```
Comment (unclegcb) [inline fix-discourse-usernames.php:50]: I'm having trouble imagining a case where this would happen based on the way your query is structured, but I guess there's no harm in some defensive code in a repair script.

File: fix-discourse-usernames.php
Comment (unclegcb) [inline fix-discourse-usernames.php:1]: How does this drush script get run? Is it just a php script you are running using drush? If we see it as something that could be useful in the future (I think so!) I'd love to give it the rest of the real drush infrastructure.

**Review by unclegcb — APPROVED** [review]
This code all looks good to me -- it seems like we have settled on keeping the usernames in sync but allowing them to be ugly? Am I reading that correctly, that since Discourse is difficult about the username that they are getting created with the ugly username and we are then pulling that back to Drupal?

I'm definitely curious to get Bill's feedback on how it's working for his team. Perhaps we need to change something in the creation process to prevent the establishment of these terrible usernames in a separate ticket.

File: fix-discourse-usernames.php
Comment (aaronbauman) [reply fix-discourse-usernames.php:1]: yeah, quick and dirty script intended for one-time use to clean up the existing disconnect.

**Comment by aaronbauman** [comment]
> it seems like we have settled on keeping the usernames in sync but allowing them to be ugly? Am I reading that correctly, that since Discourse is difficult about the username that they are getting created with the ugly username and we are then pulling that back to Drupal?

tl; dr: the Discourse user will have an ugly username until it gets changed in Drupal, and the user SSO's from Drupal to Discourse.

The username gets created in Drupal, and subsequently Discourse. That's fine and working great, so long as the Drupal username never changes.

When the Drupal username changes, we have a couple issues to reconcile:
1. Discourse doesn't expose an API endpoint to change Discourse username, so we can't use e.g. `hook_user_update` to keep it in sync
2. Discourse API uses the _Discourse username_ as a primary identifier
3. Discourse SSO uses the _Drupal uid_ for an identifier, so we're fine there
4. Discourse username (the canonical, remote one) gets updated during SSO handshake, and only during SSO handshake. This happens within the `discourse_sso` module.
5. `discourse_membership` module never records this change to `$account->discourse_user_field`

Given all that, we can leverage `hook_discourse_sso_parameters_alter` to update Drupal's record of the Discourse username, in order to keep Drupal is in sync with Discourse when the Discourse username is updated via SSO.

So, yes, we should could also figure out a better auto-generated username. But we still need a solution like this one to keep our Drupal-side Discourse data in sync.

I couldn't find an explicit explanation of Discourse's SSO behavior, so I came upon this solution after reading through various Discourse discussions and some trial and error.

Here's a thread about the Discourse setting which controls how the username can be changed:
https://meta.discourse.org/t/ldap-plugin-always-auth-overrides-username/213908

Here's the SSO doc, which includes reference for updating the username:
https://meta.discourse.org/t/setup-discourseconnect-official-single-sign-on-for-discourse-sso/13045

**Comment by unclegcb** [comment]
Thanks for the breakdown! What a rat's nest.
